language/move-stdlib/src/natives/string.rs

```diff
@@ -0,0 +1,98 @@
+// Copyright (c) The Diem Core Contributors
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+//! Implementation of native functions for utf8 strings.
+
+use move_binary_format::errors::PartialVMResult;
+use move_vm_runtime::native_functions::NativeContext;
+use move_vm_types::{
+    gas_schedule::NativeCostIndex,
+    loaded_data::runtime_types::Type,
+    natives::function::{native_gas, NativeResult},
+    pop_arg,
+    values::{Value, VectorRef},
+};
+
+use std::collections::VecDeque;
+
+// The implementation approach delegates all utf8 handling to Rust.
+// This is possible without copying of bytes because (a) we can
+// get a `std::cell::Ref<Vec<u8>>` from a `vector<u8>` and in turn a `&[u8]`
+// from that (b) assuming that `vector<u8>` embedded in a string
+// is already valid utf8, we can use `str::from_utf8_unchecked` to
+// create a `&str` view on the bytes without a copy. Once we have this
+// view, we can call ut8 functions like length, substring, etc.
+
+pub fn native_check_utf8(
+    context: &mut NativeContext,
+    _ty_args: Vec<Type>,
+    mut args: VecDeque<Value>,
+) -> PartialVMResult<NativeResult> {
+    debug_assert!(args.len() == 1);
+    let s_arg = pop_arg!(args, VectorRef);
+    let s_ref = s_arg.as_bytes_ref();
+    let ok = std::str::from_utf8(s_ref.as_slice()).is_ok();
+    // TODO: extensible native cost tables
+    let cost = native_gas(context.cost_table(), NativeCostIndex::PUSH_BACK, 1);
```

Comment:
Can we just add new indices for the string ops here?

Reply:
@vgao1996 is currently working on extending and consolidating the gas model. The problem here is that we can extend the indices for natives within the Move repo, but not outside.